### ISSUE TITLE
Tracking Issue: Towards a more redis-compatible database

### ISSUE THREAD

**Issue by PragmaTwice** [apache/kvrocks#1942]
In order to be more compatible with redis, here are commands that exist in Redis (excluding Redis Stack) but are not supported in Kvrocks:

- [x] #1944
- [x] #1945
- [x] #1946
- [x] #1947
- [x] #1948
- [ ] #1949
- [x] #1950
- [x] #1951
- [x] #1952
- [x] #1954
- [x] #1955
- [x] #1958
- [x] #1959
- [x] #1983
- [x] #1984
- [x] #1985
- [x] #2166 
- [x] #2168
- [x] #2185
- [ ] ... (maybe not fully listed)

We very much welcome anyone to claim the tasks and start to contribute to Kvrocks. If you need any help during the contribution process, you can leave a message under the corresponding issue.

Also, the parts of existing commands that are incompatible with redis are worthy of attention and welcome any contribution.

- [x] #1212
- [x] #2006
- [x] #2014
- [ ] ... adding more

It succeed #595.

**Comment by tisonkun** [comment]
Test suite in other client impl can help.

We writes C++ unit tests and Golang cases with redis-go. IIRC redis-go runs regression tests for Kvrocks also.

I tried Redisson but still have some troubles - https://github.com/apache/kvrocks/discussions/1912

Perhaps also other clients in https://redis.io/resources/clients/

**Comment by jihuayu** [comment]
We can add support for `BITOPS` command BIT/BYTE option 
- https://github.com/apache/kvrocks/issues/1212

**Comment by git-hulk** [comment]
@jihuayu done

**Comment by judahrand** [comment]
One command which would be **_amazing_** to have would be: `CLIENT TRACKING` (https://redis.io/docs/latest/commands/client-tracking/). `redis-py` just added support for easily setting up a cache which is automatically invalidated by Redis and it would be incredible if KvRocks could do this too.

**Comment by PragmaTwice** [comment]
> One command which would be **_amazing_** to have would be: `CLIENT TRACKING` (https://redis.io/docs/latest/commands/client-tracking/). `redis-py` just added support for easily setting up a cache which is automatically invalidated by Redis and it would be incredible if KvRocks could do this too.

Yeah client-side caching is in our roadmap. cc @git-hulk

**Comment by git-hulk** [comment]
Yes, one obstacle we need to overcome is knowing if a key has expired.

**Comment by PragmaTwice** [comment]
> Yes, one obstacle we need to overcome is knowing if a key has expired.

For me I think we can move the responsibility of expiration detection from the server to clients. i.e. clients should store the key-value cache as well as the TTL.

**Comment by git-hulk** [comment]
> > Yes, one obstacle we need to overcome is knowing if a key has expired.
> 
> For me I think we can move the responsibility of expiration detection from the server to clients. i.e. clients should store the key-value cache as well as the TTL.

For Redis client-side caching, it will send an invalid message notification once it's expired or evicted due to the memory policy. I'm not sure whether most SDKs will add a default TTL for keys or not. If yes, it should be fine. Or it should be a problem since the expiration needs to depend on the compaction timing.

**Comment by PragmaTwice** [comment]
> > > Yes, one obstacle we need to overcome is knowing if a key has expired.
> > 
> > 
> > For me I think we can move the responsibility of expiration detection from the server to clients. i.e. clients should store the key-value cache as well as the TTL.
> 
> For Redis client-side caching, it will send an invalid message notification once it's expired or evicted due to the memory policy. I'm not sure whether most SDKs will add a default TTL for keys or not. If yes, it should be fine. Or it should be a problem since the expiration needs to depend on the compaction timing.

It's hard for kvrocks to send notifications on expiration, but for any other writing we can notify, of course. We need to state this point to client implementations to let them store the TTL.

It's also recommended in the Redis documentation for client implementation hints:
> Handling TTLs: make sure you also request the key TTL and set the TTL in the local cache if you want to support caching keys with a TTL.

**Comment by git-hulk** [comment]
> It's also recommended in the Redis documentation for client implementation hints:

Yes, that makes sense. I will go through the Redis server and its SDK implementation for client-side caching when I get time. And then propose a discussion or design.

**Comment by judahrand** [comment]
Was there ever any progress on the server assisted client side caching?
